Src/FluentAssertions/Collections/GenericCollectionAssertions.cs

```diff
@@ -2671,6 +2671,64 @@ public AndConstraint<TAssertions> OnlyHaveUniqueItems(string because = "", param
             return new AndConstraint<TAssertions>((TAssertions)this);
         }
 
+        /// <summary>
+        /// Asserts that a collection contains only items which meet
+        /// the criteria provided by the inspector.
+        /// </summary>
+        /// <param name="expected">
+        /// The element inspector, which inspects each element in turn.
+        /// </param>
+        /// <param name="because">
+        /// A formatted phrase as is supported by <see cref="string.Format(string,object[])" /> explaining why the assertion
+        /// is needed. If the phrase does not start with the word <i>because</i>, it is prepended automatically.
+        /// </param>
+        /// <param name="becauseArgs">
+        /// Zero or more objects to format using the placeholders in <paramref name="because"/>.
+        /// </param>
+        /// <exception cref="ArgumentNullException"><paramref name="expected"/> is <c>null</c>.</exception>
+        public AndConstraint<TAssertions> AllSatisfy(Action<T> expected, string because = "", params object[] becauseArgs)
+        {
+            Guard.ThrowIfArgumentIsNull(expected, nameof(expected), "Cannot verify against a <null> inspector");
+
+            var success = Execute.Assertion
+                                 .BecauseOf(because, becauseArgs)
+                                 .WithExpectation(
+                                     "Expected {context:collection} to contain only items satisfying the inspector{reason}, ")
+                                 .Given(() => Subject)
+                                 .ForCondition(subject => subject is not null)
+                                 .FailWith("but collection is <null>.")
+                                 .Then
+                                 .ForCondition(subject => subject.Any())
+                                 .FailWith("but collection is empty.")
+                                 .Then
+                                 .ClearExpectation();
+
+            if (success != false) return new AndConstraint<TAssertions>((TAssertions)this);
```

Comment:
`success` is always false (internally, `succeeded` is `null`) at this point with the code above, seemingly due to `FailWith` setting `succeeded` to null at https://github.com/fluentassertions/fluentassertions/blob/81368ad859f4b870c7c5d51b35084dca01cc4ae9/Src/FluentAssertions/Execution/AssertionScope.cs#L278-L281 Is there a step I am missing? It seems like `success` will always be false if called after `FailWith` followed by `ClearExpectation`

Reply:
No, it returns a `Continuation` instance which has the real result. And this continuation can be implicitly converted to `bool`.

Reply:
I think there is a miscommunication on my end here. When implicitly converting to a bool, the result of that conversion is always false in the code above, regardless of the outcomes from the preceding `ForCondition` and `FailWith` calls. If I change the code to check for success:
```csharp
bool success = Execute.Assertion
    .BecauseOf(because, becauseArgs)
    .WithExpectation("Expected {context:collection} to contain only items satisfying the inspector{reason}, ")
    .Given(() => Subject)
    .ForCondition(subject => subject is not null)
    .FailWith("but collection is <null>.")
    .Then
    .ForCondition(subject => subject.Any())
    .FailWith("but collection is empty.")
    .Then
    .ClearExpectation();

if (!success)
{
    return new AndConstraint<TAssertions>((TAssertions)this);
}
```
the test currently named `When_collection_asserting_with_any_item_that_does_not_satisfy_inspector_it_should_throw` fails because `success` is false and the code exits early, never evaluating the inspector against the collection items. 

I don't understand the internals but it seems like any call to `ClearExpectation` that immediately follows a `FailWith` will subsequently implicitly convert to `false` due to the `FailWith` in `AssertionScope` having a finally block that sets `succeeded` to null 

https://github.com/fluentassertions/fluentassertions/blob/81368ad859f4b870c7c5d51b35084dca01cc4ae9/Src/FluentAssertions/Execution/AssertionScope.cs#L278-L281

Subsequently, `predecessor.succeeded` is `null` after the call to `FailWith` on line 93 in this section of code: 
https://github.com/fluentassertions/fluentassertions/blob/81368ad859f4b870c7c5d51b35084dca01cc4ae9/Src/FluentAssertions/Execution/GivenSelector.cs#L89-L98

`predecesor.succeeded` is what is used in the implicit conversion to `bool` due to `Then.ClearExpectation`:

https://github.com/fluentassertions/fluentassertions/blob/81368ad859f4b870c7c5d51b35084dca01cc4ae9/Src/FluentAssertions/Execution/ContinuationOfGiven.cs#L20-L28

https://github.com/fluentassertions/fluentassertions/blob/81368ad859f4b870c7c5d51b35084dca01cc4ae9/Src/FluentAssertions/Execution/GivenSelector.cs#L101-L105

This same situation can be observed by modifying the `SatisfyRespectively` method from:

```csharp
bool success = Execute.Assertion
    .BecauseOf(because, becauseArgs)
    .WithExpectation("Expected {context:collection} to satisfy all inspectors{reason}, ")
    .Given(() => Subject)
    .ForCondition(subject => subject is not null)
    .FailWith("but collection is <null>.")
    .Then
    .ForCondition(subject => subject.Any())
    .FailWith("but collection is empty.")
    .Then
    .ClearExpectation()
    .Then
    .Given(subject => (elements: subject.Count(), inspectors: elementInspectors.Count))
    .ForCondition(count => count.elements == count.inspectors)
    .FailWith(
        "Expected {context:collection} to contain exactly {0} items{reason}, but it contains {1} items",
        count => count.inspectors, count => count.elements);
```
to
```csharp
bool success = Execute.Assertion
    .BecauseOf(because, becauseArgs)
    .WithExpectation("Expected {context:collection} to satisfy all inspectors{reason}, ")
    .Given(() => Subject)
    .ForCondition(subject => subject is not null)
    .FailWith("but collection is <null>.")
    .Then
    .ForCondition(subject => subject.Any())
    .FailWith("but collection is empty.")
    .Then
    .ClearExpectation();
```
With the modified code, the test `When_asserting_collection_does_not_satisfy_any_inspector_it_should_throw` will fail because the code exits early due to `success` evaluating to false.

The implementation of `Satisfy` below has the same sort of code written without calls to `WithExpectation` or `ClearExpectation`:
```csharp
bool success = Execute.Assertion
    .BecauseOf(because, becauseArgs)
    .Given(() => Subject)
    .ForCondition(subject => subject is not null)
    .FailWith("Expected {context:collection} to satisfy all predicates{reason}, but collection is <null>.")
    .Then
    .ForCondition(subject => subject.Any())
    .FailWith("Expected {context:collection} to satisfy all predicates{reason}, but collection is empty.");

if (success)
{
    // Never hit
}
```
Do I need to rewrite `AllSatisfy` in a way that avoids `WithExpectation` and `ClearExpectation`?

Reply:
Maybe related to #1596

Reply:
I'll take a look.

Reply:
I think I've fixed it. See https://github.com/fluentassertions/fluentassertions/pull/1791

Reply:
Please rebase on `master` and then it'll work as I suggested in the beginning of this thread.